File: sdk/go/k8s.go
package mitos

// Minimal Kubernetes REST client for cluster mode. The Go SDK stays
// dependency-free: direct mode (SandboxServer) is untouched and pulls nothing,
// and cluster mode here is implemented on the Go standard library alone
// (net/http, crypto/tls, crypto/x509, encoding/json, encoding/base64). It does
// NOT pull k8s.io/client-go, which is large and would drag a transitive
// dependency tree into every consumer of this module.
//
// The client speaks the custom-resource REST paths directly:
//
//	/apis/mitos.run/v1/namespaces/{ns}/{plural}            (list, create)
//	/apis/mitos.run/v1/namespaces/{ns}/{plural}/{name}     (get, delete, patch)
//
// and reads core Secrets at /api/v1/namespaces/{ns}/secrets/{name}. Auth is a
// bearer token; TLS trusts the cluster CA. The token VALUE is never logged.

import (
	"bytes"
	"context"
	"crypto/tls"
	"crypto/x509"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net"
	"net/http"
	"os"
	"path/filepath"
	"strings"
)

// k8sAPIGroup, k8sAPIVersion are the CRD group/version the cluster client drives.
const (
	k8sAPIGroup   = "mitos.run"
	k8sAPIVersion = "v1"
)

// In-cluster service-account mount paths (the standard Kubernetes locations).
const (
	inClusterTokenPath = "/var/run/secrets/kubernetes.io/serviceaccount/token" //nolint:gosec // path, not a secret value
	inClusterCAPath    = "/var/run/secrets/kubernetes.io/serviceaccount/ca.crt"
)

// k8sConfig is the resolved connection to the API server: the server URL, the
// bearer token, and an HTTP client whose TLS trusts the cluster CA. The token
// VALUE is held in memory only and is never logged.
type k8sConfig struct {
	server string
	token  string
	http   *http.Client
}

// k8sObject is a Kubernetes custom object as a generic JSON map, mirroring the
// untyped dict the Python and TypeScript SDKs pass around. Helpers below read
// nested spec/status/metadata fields defensively (a missing key yields a zero
// value, never a panic), matching the Python SDK's obj.get(...).
type k8sObject map[string]any

// k8sObjectList is the wire shape of a CRD list response.
type k8sObjectList struct {
	Items []k8sObject `json:"items"`
}

// loadInClusterConfig builds a k8sConfig from the in-cluster service-account
// mount: the API server from KUBERNETES_SERVICE_HOST/PORT, the bearer token
// from the projected token file, and the CA from the mounted ca.crt. It returns
// an actionable AgentRunError when run outside a cluster (the env vars are
// absent) so the failure names the fix rather than surfacing a bare file error.
func loadInClusterConfig() (*k8sConfig, error) {
	host := os.Getenv("KUBERNETES_SERVICE_HOST")
	port := os.Getenv("KUBERNETES_SERVICE_PORT")
	if host == "" || port == "" {
		return nil, &Error{
			Code:        "not_in_cluster",
			Message:     "in-cluster config requested but KUBERNETES_SERVICE_HOST/PORT are not set",
			Cause:       "the in-cluster service-account environment is not present",
			Remediation: "Run inside a Kubernetes pod, or use a kubeconfig with WithKubeconfig(path).",
		}
	}
	tokenBytes, err := os.ReadFile(inClusterTokenPath)
	if err != nil {
		return nil, &Error{
			Code:        "incluster_token_unreadable",
			Message:     "could not read the in-cluster service-account token",
			Cause:       err.Error(),
			Remediation: "Ensure the pod mounts a service-account token at " + inClusterTokenPath + ".",
		}
	}
	caBytes, err := os.ReadFile(inClusterCAPath)
	if err != nil {
		return nil, &Error{
			Code:        "incluster_ca_unreadable",
			Message:     "could not read the in-cluster CA certificate",
			Cause:       err.Error(),
			Remediation: "Ensure the pod mounts the cluster CA at " + inClusterCAPath + ".",
		}
	}
	httpClient, err := httpClientForCA(caBytes)
	if err != nil {
		return nil, err
	}
	server := "https://" + net.JoinHostPort(host, port)
	return &k8sConfig{
		server: server,
		token:  strings.TrimSpace(string(tokenBytes)),
		http:   httpClient,
	}, nil
}

// httpClientForCA builds an *http.Client whose TLS root pool trusts caPEM. When
// caPEM is empty the client uses the system roots (a kubeconfig may carry no
// inline CA). A CA that fails to parse is a typed error so the misconfiguration
// is legible.
func httpClientForCA(caPEM []byte) (*http.Client, error) {
	tlsConfig := &tls.Config{MinVersion: tls.VersionTLS12}
	if len(caPEM) > 0 {
		pool := x509.NewCertPool()
		if !pool.AppendCertsFromPEM(caPEM) {
			return nil, &Error{
				Code:        "ca_parse_failed",
				Message:     "the cluster CA certificate could not be parsed",
				Cause:       "AppendCertsFromPEM rejected the CA PEM",
				Remediation: "Check the CA certificate is valid PEM (the kubeconfig certificate-authority-data or the mounted ca.crt).",
			}
		}
		tlsConfig.RootCAs = pool
	}
	return &http.Client{
		Transport: &http.Transport{TLSClientConfig: tlsConfig},
	}, nil
}

// k8sClient drives the CRD and Secret REST endpoints for one namespace's worth
// of cluster-mode calls. It is the Go analogue of the Python CustomObjectsApi +
// CoreV1Api pair and the TypeScript K8sApi seam.
type k8sClient struct {
	cfg *k8sConfig
}

// newK8sClient builds a client over a resolved config.
func newK8sClient(cfg *k8sConfig) *k8sClient {
	return &k8sClient{cfg: cfg}
}

// crdPath builds the REST path for a namespaced custom resource. When name is
// empty it is the collection path (list, create); otherwise the item path (get,
// delete, patch).
func crdPath(namespace, plural, name string) string {
	base := fmt.Sprintf("/apis/%s/%s/namespaces/%s/%s", k8sAPIGroup, k8sAPIVersion, namespace, plural)
	if name == "" {
		return base
	}
	return base + "/" + name
}

// secretPath builds the core Secret item path.
func secretPath(namespace, name string) string {
	return fmt.Sprintf("/api/v1/namespaces/%s/secrets/%s", namespace, name)
}

// getObject GETs a single custom object. A 404 surfaces as an *Error with
// Status 404 so callers (the lazy default-pool path, reconnect-by-name) can tell
// absent from a real failure.
func (c *k8sClient) getObject(ctx context.Context, namespace, plural, name string) (k8sObject, error) {
	var obj k8sObject
	if err := c.do(ctx, http.MethodGet, crdPath(namespace, plural, name), "", nil, &obj); err != nil {
		return nil, err
	}
	return obj, nil
}

// listObjects GETs the collection of custom objects.
func (c *k8sClient) listObjects(ctx context.Context, namespace, plural string) (*k8sObjectList, error) {
	var list k8sObjectList
	if err := c.do(ctx, http.MethodGet, crdPath(namespace, plural, ""), "", nil, &list); err != nil {
		return nil, err
	}
	return &list, nil
}

// createObject POSTs a custom object to the collection.
func (c *k8sClient) createObject(ctx context.Context, namespace, plural string, body k8sObject) (k8sObject, error) {
	var out k8sObject
	if err := c.do(ctx, http.MethodPost, crdPath(namespace, plural, ""), "application/json", body, &out); err != nil {
		return nil, err
	}
	return out, nil
}

// deleteObject DELETEs a single custom object.
func (c *k8sClient) deleteObject(ctx context.Context, namespace, plural, name string) error {
	return c.do(ctx, http.MethodDelete, crdPath(namespace, plural, name), "", nil, nil)
}

// readSecret reads a core Secret and returns its data decoded to UTF-8 strings
// keyed by the Secret key. A 404 surfaces as an *Error with Status 404 so a
// caller can tolerate a missing token Secret. Secret VALUES are held in memory
// only and are NEVER logged.
func (c *k8sClient) readSecret(ctx context.Context, namespace, name string) (map[string]string, error) {
	var obj struct {
		Data map[string]string `json:"data"`
	}
	if err := c.do(ctx, http.MethodGet, secretPath(namespace, name), "", nil, &obj); err != nil {
		return nil, err
	}
	out := make(map[string]string, len(obj.Data))
	for k, v := range obj.Data {
		decoded, err := base64.StdEncoding.DecodeString(v)
		if err != nil {
			// A non-base64 Secret value is unexpected; skip it rather than
			// surfacing the raw value (which could be a secret) in an error.
			continue
		}
		out[k] = string(decoded)
	}
	return out, nil
}

// do issues a Kubernetes API request with the bearer token and decodes a 2xx
// JSON body into out (when non-nil). A non-2xx response is parsed into a typed
// *Error carrying the Kubernetes Status reason/code; the bearer token is never
// placed in an error. contentType, when set, is sent for a body-carrying request
// (application/json for create, application/merge-patch+json for patch).
func (c *k8sClient) do(ctx context.Context, method, path, contentType string, body any, out any) error {
	var reqBody io.Reader
	if body != nil {
		buf, err := json.Marshal(body)
		if err != nil {
			return fmt.Errorf("encode request body: %w", err)
		}
		reqBody = bytes.NewReader(buf)
	}
	req, err := http.NewRequestWithContext(ctx, method, c.cfg.server+path, reqBody)
	if err != nil {
		return fmt.Errorf("build request: %w", err)
	}
	req.Header.Set("Accept", "application/json")
	if contentType != "" {
		req.Header.Set("Content-Type", contentType)
	}
	if c.cfg.token != "" {
		req.Header.Set("Authorization", "Bearer "+c.cfg.token)
	}

	resp, err := c.cfg.http.Do(req)
	if err != nil {
		return fmt.Errorf("request %s %s: %w", method, path, err)
	}
	defer func() { _ = resp.Body.Close() }()

	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return fmt.Errorf("read response: %w", err)
	}
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return parseK8sError(resp.StatusCode, respBody)
	}
	if out != nil && len(respBody) > 0 {
		if err := json.Unmarshal(respBody, out); err != nil {
			return fmt.Errorf("decode response: %w", err)
		}
	}
	return nil
}

// k8sStatus is the Kubernetes Status object returned on an API error.
type k8sStatus struct {
	Message string `json:"message"`
	Reason  string `json:"reason"`
	Code    int    `json:"code"`
}

// parseK8sError turns a non-2xx Kubernetes response into a typed *Error. The
// HTTP status is preserved on Error.Status so callers can branch on 404 (absent)
// versus 409 (already exists). The token never appears in a Status body, so no
// redaction is needed here.
func parseK8sError(status int, body []byte) error {
	var st k8sStatus
	msg := strings.TrimSpace(string(body))
	if err := json.Unmarshal(body, &st); err == nil && (st.Message != "" || st.Reason != "") {
		message := st.Message
		if message == "" {
			message = st.Reason
		}
		return &Error{
			Code:        "k8s_" + k8sErrorCode(status, st.Reason),
			Message:     message,
			Cause:       fmt.Sprintf("kubernetes API returned %d %s", status, st.Reason),
			Remediation: k8sRemediation(status),
			Status:      status,
		}
	}
	return &Error{
		Code:    "k8s_http_error",
		Message: fmt.Sprintf("kubernetes API returned HTTP %d", status),
		Cause:   msg,
		Status:  status,
	}
}

// k8sErrorCode maps an HTTP status to a short, stable code suffix.
func k8sErrorCode(status int, reason string) string {
	switch status {
	case http.StatusNotFound:
		return "not_found"
	case http.StatusConflict:
		return "conflict"
	case http.StatusForbidden:
		return "forbidden"
	case http.StatusUnauthorized:
		return "unauthorized"
	}
	if reason != "" {
		return strings.ToLower(reason)
	}
	return "error"
}

// k8sRemediation returns an actionable hint for common API errors.
func k8sRemediation(status int) string {
	switch status {
	case http.StatusNotFound:
		return "Check the object name and namespace; create it first if it does not exist."
	case http.StatusForbidden, http.StatusUnauthorized:
		return "Check the service-account RBAC grants access to mitos.run resources in this namespace."
	}
	return ""
}

// statusOf returns the HTTP status carried by an *Error, or 0 when err is not an
// *Error. It lets the cluster logic test for 404/409 without string matching.
func statusOf(err error) int {
	var e *Error
	if errors.As(err, &e) {
		return e.Status
	}
	return 0
}

// kubeconfig is the minimal subset of a kubeconfig file the SDK parses. Only the
// current-context's cluster (server + CA) and user (token, or client cert/key)
// are read; the rest of the file is ignored. See loadKubeconfig for the
// supported subset.
type kubeconfig struct {
	CurrentContext string `json:"current-context" yaml:"current-context"`
	Clusters       []struct {
		Name    string `json:"name" yaml:"name"`
		Cluster struct {
			Server                   string `json:"server" yaml:"server"`
			CertificateAuthorityData string `json:"certificate-authority-data" yaml:"certificate-authority-data"`
			CertificateAuthority     string `json:"certificate-authority" yaml:"certificate-authority"`
			InsecureSkipTLSVerify    bool   `json:"insecure-skip-tls-verify" yaml:"insecure-skip-tls-verify"`
		} `json:"cluster" yaml:"cluster"`
	} `json:"clusters" yaml:"clusters"`
	Contexts []struct {
		Name    string `json:"name" yaml:"name"`
		Context struct {
			Cluster   string `json:"cluster" yaml:"cluster"`
			User      string `json:"user" yaml:"user"`
			Namespace string `json:"namespace" yaml:"namespace"`
		} `json:"context" yaml:"context"`
	} `json:"contexts" yaml:"contexts"`
	Users []struct {
		Name string `json:"name" yaml:"name"`
		User struct {
			Token                 string `json:"token" yaml:"token"`
			ClientCertificateData string `json:"client-certificate-data" yaml:"client-certificate-data"`
			ClientKeyData         string `json:"client-key-data" yaml:"client-key-data"`
		} `json:"user" yaml:"user"`
	} `json:"users" yaml:"users"`
}

// loadKubeconfig parses a kubeconfig file and resolves the current context into
// a k8sConfig. It supports the common subset:
//
//   - server URL and inline certificate-authority-data (base64 PEM), or a
//     certificate-authority file path, or system roots when neither is set;
//   - user auth by bearer token, or by inline client-certificate-data +
//     client-key-data (base64 PEM) for mutual-TLS clusters (kind, minikube).
//
// It does NOT support exec credential plugins, auth-provider blocks, or external
// command tokens. The file is parsed as YAML via a minimal embedded decoder
// (yamlToJSON) so the SDK keeps its zero third-party dependencies. The path
// defaults to $KUBECONFIG, then $HOME/.kube/config, when empty.
func loadKubeconfig(path string) (*k8sConfig, error) {
	if path == "" {
		path = os.Getenv("KUBECONFIG")
	}
	if path == "" {
		home, err := os.UserHomeDir()
		if err == nil && home != "" {
			path = filepath.Join(home, ".kube", "config")
		}
	}
	if path == "" {
		return nil, &Error{
			Code:        "kubeconfig_not_found",
			Message:     "no kubeconfig path and no $HOME to derive ~/.kube/config",
			Cause:       "neither WithKubeconfig nor $KUBECONFIG nor $HOME is set",
			Remediation: "Pass WithKubeconfig(path), set $KUBECONFIG, or use WithInCluster() inside a pod.",
		}
	}
	raw, err := os.ReadFile(path) //nolint:gosec // path supplied by the caller / their KUBECONFIG
	if err != nil {
		return nil, &Error{
			Code:        "kubeconfig_unreadable",
			Message:     "could not read the kubeconfig file",
			Cause:       err.Error(),
			Remediation: "Check the kubeconfig path is correct and readable: " + path,
		}
	}
	jsonBytes, err := yamlToJSON(raw)
	if err != nil {
		return nil, &Error{
			Code:        "kubeconfig_parse_failed",
			Message:     "could not parse the kubeconfig file",
			Cause:       err.Error(),
			Remediation: "The SDK parses a common kubeconfig subset (no exec/auth-provider plugins); check the file or run inside the cluster with WithInCluster().",
		}
	}
	var kc kubeconfig
	if err := json.Unmarshal(jsonBytes, &kc); err != nil {
		return nil, &Error{
			Code:        "kubeconfig_parse_failed",
			Message:     "could not decode the kubeconfig structure",
			Cause:       err.Error(),
			Remediation: "Check the kubeconfig has clusters, contexts, and users entries.",
		}
	}
	return kc.resolve(filepath.Dir(path))
}

// resolve turns a parsed kubeconfig into a connection for its current context.
// baseDir is the kubeconfig's directory, used to resolve a relative
// certificate-authority file path.
func (kc *kubeconfig) resolve(baseDir string) (*k8sConfig, error) {
	if kc.CurrentContext == "" {
		return nil, &Error{
			Code:        "kubeconfig_no_context",
			Message:     "the kubeconfig has no current-context",
			Cause:       "current-context is empty",
			Remediation: "Set a current context with kubectl config use-context <name>.",
		}
	}
	var clusterName, userName string
	for _, ctx := range kc.Contexts {
		if ctx.Name == kc.CurrentContext {
			clusterName = ctx.Context.Cluster
			userName = ctx.Context.User
			break
		}
	}
	if clusterName == "" {
		return nil, &Error{
			Code:        "kubeconfig_context_missing",
			Message:     "the current-context is not defined in the kubeconfig",
			Cause:       "no contexts entry matches current-context " + kc.CurrentContext,
			Remediation: "Check the kubeconfig contexts list includes " + kc.CurrentContext + ".",
		}
	}

	var server string
	var caPEM []byte
	for _, cl := range kc.Clusters {
		if cl.Name != clusterName {
			continue
		}
		server = cl.Cluster.Server
		switch {
		case cl.Cluster.CertificateAuthorityData != "":
			decoded, err := base64.StdEncoding.DecodeString(cl.Cluster.CertificateAuthorityData)
			if err != nil {
				return nil, &Error{
					Code:        "kubeconfig_ca_invalid",
					Message:     "certificate-authority-data is not valid base64",
					Cause:       err.Error(),
					Remediation: "Regenerate the kubeconfig or check the certificate-authority-data field.",
				}
			}
			caPEM = decoded
		case cl.Cluster.CertificateAuthority != "":
			p := cl.Cluster.CertificateAuthority
			if !filepath.IsAbs(p) {
				p = filepath.Join(baseDir, p)
			}
			b, err := os.ReadFile(p) //nolint:gosec // path from the user's own kubeconfig
			if err != nil {
				return nil, &Error{
					Code:        "kubeconfig_ca_unreadable",
					Message:     "could not read the certificate-authority file",
					Cause:       err.Error(),
					Remediation: "Check the certificate-authority path in the kubeconfig: " + p,
				}
			}
			caPEM = b
		}
		break
	}
	if server == "" {
		return nil, &Error{
			Code:        "kubeconfig_no_server",
			Message:     "the current cluster has no server URL",
			Cause:       "cluster " + clusterName + " has an empty server field",
			Remediation: "Check the kubeconfig clusters entry for " + clusterName + ".",
		}
	}

	httpClient, err := httpClientForCA(caPEM)
	if err != nil {
		return nil, err
	}

	var token string
	var clientCertPEM, clientKeyPEM []byte
	for _, u := range kc.Users {
		if u.Name != userName {
			continue
		}
		token = u.User.Token
		if u.User.ClientCertificateData != "" && u.User.ClientKeyData != "" {
			cert, errC := base64.StdEncoding.DecodeString(u.User.ClientCertificateData)
			key, errK := base64.StdEncoding.DecodeString(u.User.ClientKeyData)
			if errC == nil && errK == nil {
				clientCertPEM = cert
				clientKeyPEM = key
			}
		}
		break
	}

	if len(clientCertPEM) > 0 && len(clientKeyPEM) > 0 {
		cert, err := tls.X509KeyPair(clientCertPEM, clientKeyPEM)
		if err != nil {
			return nil, &Error{
				Code:        "kubeconfig_client_cert_invalid",
				Message:     "the kubeconfig client certificate or key is invalid",
				Cause:       err.Error(),
				Remediation: "Check client-certificate-data and client-key-data in the kubeconfig.",
			}
		}
		if tr, ok := httpClient.Transport.(*http.Transport); ok {
			tr.TLSClientConfig.Certificates = []tls.Certificate{cert}
		}
	}

	return &k8sConfig{server: strings.TrimRight(server, "/"), token: token, http: httpClient}, nil
}

File: sdk/ruby/lib/mitos/k8s.rb
# frozen_string_literal: true

require "base64"
require "json"
require "net/http"
require "openssl"
require "uri"
require "yaml"

require "mitos/errors"

module Mitos
  # A minimal Kubernetes REST client built on the Ruby standard library only
  # (net/http for transport, openssl for TLS and the CA bundle, json for the
  # bodies, yaml for kubeconfig parsing, base64 for the in-cluster and secret
  # token decoding). It exists so the Ruby SDK's cluster mode stays dependency
  # free: there is no equivalent of the Python "kubernetes" client or the
  # TypeScript "@kubernetes/client-node" here, just direct REST against the API
  # server.
  #
  # The client speaks only the namespaced custom-resource and Secret endpoints
  # the AgentRun surface needs:
  #
  #   GET    /apis/{group}/{version}/namespaces/{ns}/{plural}
  #   GET    /apis/{group}/{version}/namespaces/{ns}/{plural}/{name}
  #   POST   /apis/{group}/{version}/namespaces/{ns}/{plural}
  #   DELETE /apis/{group}/{version}/namespaces/{ns}/{plural}/{name}
  #   GET    /api/v1/namespaces/{ns}/secrets/{name}
  #
  # Auth is a bearer token (in-cluster service-account token or a kubeconfig
  # token) or client-certificate mTLS (a kubeconfig client cert and key). The
  # token VALUE is never logged and never placed in an error message.
  class K8sClient
    # The in-cluster service-account mount, mounted into every pod by the
    # kubelet. The token rotates; it is re-read on each request so a rotated
    # token is picked up without recreating the client.
    SA_DIR = "/var/run/secrets/kubernetes.io/serviceaccount"
    SA_TOKEN = "#{SA_DIR}/token"
    SA_CA = "#{SA_DIR}/ca.crt"
    SA_NAMESPACE = "#{SA_DIR}/namespace"

    attr_reader :server, :default_namespace

    # Build the client from a resolved config Hash (see .in_cluster and
    # .from_kubeconfig). The config carries: :server (the API base URL),
    # :ca_data (PEM string or nil for the system store), :token (bearer, or
    # nil), :token_file (a path re-read per request, or nil), :client_cert_data
    # and :client_key_data (PEM strings for mTLS, or nil), :insecure (skip TLS
    # verification, only honored when a kubeconfig explicitly sets it), and
    # :namespace (the default namespace, or nil).
    def initialize(config)
      @server = config[:server].to_s.sub(%r{/+\z}, "")
      @ca_data = config[:ca_data]
      @token = config[:token]
      @token_file = config[:token_file]
      @client_cert_data = config[:client_cert_data]
      @client_key_data = config[:client_key_data]
      @insecure = config[:insecure] ? true : false
      @default_namespace = config[:namespace]
      raise config_error("the resolved Kubernetes server URL is empty") if @server.empty?
    end

    # Resolve the in-cluster configuration from the service-account mount and
    # the KUBERNETES_SERVICE_HOST / KUBERNETES_SERVICE_PORT environment the
    # kubelet injects. Raises a typed MitosError naming the missing piece when
    # not running inside a pod.
    def self.in_cluster
      host = ENV["KUBERNETES_SERVICE_HOST"]
      port = ENV["KUBERNETES_SERVICE_PORT"]
      if host.nil? || host.empty? || port.nil? || port.empty?
        raise MitosError.new(
          "not running in a Kubernetes cluster",
          code: "k8s_config_error",
          cause: "KUBERNETES_SERVICE_HOST / KUBERNETES_SERVICE_PORT are not set",
          remediation: "Run inside a pod with a service account, or pass kubeconfig: to AgentRun."
        )
      end
      unless File.file?(SA_TOKEN)
        raise MitosError.new(
          "in-cluster service-account token is not mounted",
          code: "k8s_config_error",
          cause: "#{SA_TOKEN} is absent",
          remediation: "Run inside a pod whose service account token is mounted, or pass kubeconfig:."
        )
      end
      ca = File.file?(SA_CA) ? File.read(SA_CA) : nil
      ns = File.file?(SA_NAMESPACE) ? File.read(SA_NAMESPACE).strip : nil
      new(
        server: "https://#{host}:#{port}",
        ca_data: ca,
        token_file: SA_TOKEN,
        namespace: (ns unless ns.nil? || ns.empty?)
      )
    end

    # Resolve configuration from a kubeconfig YAML file. +path+ defaults to
    # ENV['KUBECONFIG'] (first entry if it is a list) then ~/.kube/config. The
    # current-context cluster (server, CA) and user (token or client cert/key)
    # are read; relative file references inside the kubeconfig are resolved and
    # inlined. Raises a typed MitosError when the file or the named context is
    # missing.
    def self.from_kubeconfig(path = nil)
      resolved = resolve_kubeconfig_path(path)
      unless resolved && File.file?(resolved)
        raise MitosError.new(
          "kubeconfig not found",
          code: "k8s_config_error",
          cause: "no kubeconfig at #{resolved || '(unset)'}",
          remediation: "Set KUBECONFIG, pass kubeconfig: to AgentRun, or run AgentRun(in_cluster: true) in a pod."
        )
      end
      doc = YAML.safe_load(File.read(resolved), permitted_classes: [], aliases: false) || {}
      new(parse_kubeconfig(doc, File.dirname(File.expand_path(resolved))))
    end

    def self.resolve_kubeconfig_path(path)
      return path unless path.nil? || (path.respond_to?(:empty?) && path.empty?)

      env = ENV["KUBECONFIG"]
      unless env.nil? || env.empty?
        # KUBECONFIG may be a path list; the first existing entry wins, matching
        # the client-go merge order closely enough for the single-file case.
        first = env.split(File::PATH_SEPARATOR).find { |p| !p.empty? }
        return first if first
      end
      home = Dir.home
      return nil if home.nil? || home.empty?

      File.join(home, ".kube", "config")
    rescue StandardError
      nil
    end

    # Turn a parsed kubeconfig document into the config Hash initialize expects.
    # base_dir resolves relative certificate-authority / client-certificate /
    # token file references.
    def self.parse_kubeconfig(doc, base_dir)
      current = doc["current-context"]
      raise config_error_class("kubeconfig has no current-context") if current.nil? || current.to_s.empty?

      context = find_named(doc["contexts"], current)
      raise config_error_class("kubeconfig context #{current} not found") unless context

      ctx = context["context"] || {}
      cluster = find_named(doc["clusters"], ctx["cluster"]) || {}
      user = find_named(doc["users"], ctx["user"]) || {}
      c = cluster["cluster"] || {}
      u = user["user"] || {}

      {
        server: c["server"],
        ca_data: read_inline_or_file(c["certificate-authority-data"], c["certificate-authority"], base_dir),
        token: kubeconfig_token(u),
        client_cert_data: read_inline_or_file(u["client-certificate-data"], u["client-certificate"], base_dir),
        client_key_data: read_inline_or_file(u["client-key-data"], u["client-key"], base_dir),
        insecure: c["insecure-skip-tls-verify"] ? true : false,
        namespace: (ctx["namespace"] unless ctx["namespace"].nil? || ctx["namespace"].to_s.empty?)
      }
    end

    # Read a base64 "*-data" field if present, else read the referenced file
    # (relative paths resolved against the kubeconfig directory). Returns a PEM
    # / raw string or nil.
    def self.read_inline_or_file(data_b64, file_ref, base_dir)
      unless data_b64.nil? || data_b64.to_s.empty?
        return Base64.decode64(data_b64)
      end
      return nil if file_ref.nil? || file_ref.to_s.empty?

      path = File.expand_path(file_ref, base_dir)
      File.file?(path) ? File.read(path) : nil
    end

    # The kubeconfig user token: an inline "token", else a "tokenFile" read from
    # disk. Exec-plugin and auth-provider credential helpers are out of scope
    # for the dependency-free client; their absence surfaces as a 401 from the
    # API server with an actionable message rather than a silent tokenless call.
    def self.kubeconfig_token(user)
      tok = user["token"]
      return tok unless tok.nil? || tok.to_s.empty?

      tf = user["tokenFile"]
      return File.read(tf).strip if !tf.nil? && !tf.to_s.empty? && File.file?(tf)

      nil
    end

    def self.find_named(list, name)
      return nil unless list.is_a?(Array)

      list.find { |e| e.is_a?(Hash) && e["name"] == name }
    end

    def self.config_error_class(detail)
      MitosError.new(
        "invalid kubeconfig",
        code: "k8s_config_error",
        cause: detail,
        remediation: "Point KUBECONFIG / kubeconfig: at a valid config, or run AgentRun(in_cluster: true) in a pod."
      )
    end

    # GET a namespaced custom-object collection. Returns the decoded list body
    # (a Hash with an "items" array).
    def list_namespaced(group, version, namespace, plural)
      request(Net::HTTP::Get, custom_path(group, version, namespace, plural), nil)
    end

    # GET a single namespaced custom object by name. Returns the decoded body.
    def get_namespaced(group, version, namespace, plural, name)
      request(Net::HTTP::Get, custom_path(group, version, namespace, plural, name), nil)
    end

    # POST a namespaced custom object. +body+ is the object Hash. Returns the
    # decoded created body.
    def create_namespaced(group, version, namespace, plural, body)
      request(Net::HTTP::Post, custom_path(group, version, namespace, plural), body)
    end

    # DELETE a namespaced custom object by name. Returns the decoded body.
    def delete_namespaced(group, version, namespace, plural, name)
      request(Net::HTTP::Delete, custom_path(group, version, namespace, plural, name), nil)
    end

    # GET a core/v1 Secret by name. Returns the decoded Secret body (with the
    # base64 "data" map). The caller decodes only the key it needs and never
    # logs the value.
    def get_secret(namespace, name)
      path = "/api/v1/namespaces/#{esc(namespace)}/secrets/#{esc(name)}"
      request(Net::HTTP::Get, path, nil)
    end

    # The HTTP status of the last raised ApiError, exposed so callers can branch
    # on 404 / 409 without parsing a message. Set on the exception, not here.

    private

    def custom_path(group, version, namespace, plural, name = nil)
      base = "/apis/#{esc(group)}/#{esc(version)}/namespaces/#{esc(namespace)}/#{esc(plural)}"
      name.nil? || name.empty? ? base : "#{base}/#{esc(name)}"
    end

    def esc(segment)
      URI.encode_www_form_component(segment.to_s)
    end

    # Perform the request: build the URI, attach auth (bearer or mTLS), set the
    # TLS verification mode and CA, send the JSON body, and decode the JSON
    # response. A non-2xx status raises a typed ApiError carrying the parsed
    # Kubernetes Status (reason + code) so callers branch on .status, never the
    # message. Auth material is never logged.
    def request(method_class, path, body)
      uri = URI.parse(@server + path)
      req = method_class.new(uri.request_uri)
      req["Accept"] = "application/json"
      if body
        req["Content-Type"] = "application/json"
        req.body = JSON.generate(body)
      end
      apply_auth(req)

      http = build_http(uri)
      resp = http.request(req)
      status = resp.code.to_i
      raise api_error(status, resp.body) unless status >= 200 && status < 300

      text = resp.body
      return {} if text.nil? || text.empty?

      JSON.parse(text)
    end

    # Attach the bearer token. The token file is re-read per request so a
    # rotated in-cluster token is honored. mTLS users carry no bearer; the cert
    # is attached on the transport in build_http.
    def apply_auth(req)
      token = current_token
      req["Authorization"] = "Bearer #{token}" if token && !token.empty?
    end

    def current_token
      return @token if @token && !@token.empty?
      return nil if @token_file.nil? || @token_file.empty?

      File.read(@token_file).strip
    rescue StandardError
      nil
    end

    def build_http(uri)
      http = Net::HTTP.new(uri.host, uri.port)
      http.use_ssl = (uri.scheme == "https")
      return http unless uri.scheme == "https"

      if @insecure
        http.verify_mode = OpenSSL::SSL::VERIFY_NONE
      else
        http.verify_mode = OpenSSL::SSL::VERIFY_PEER
        if @ca_data && !@ca_data.empty?
          store = OpenSSL::X509::Store.new
          store.add_cert(OpenSSL::X509::Certificate.new(@ca_data))
          http.cert_store = store
        end
      end
      if @client_cert_data && !@client_cert_data.empty? && @client_key_data && !@client_key_data.empty?
        http.cert = OpenSSL::X509::Certificate.new(@client_cert_data)
        http.key = OpenSSL::PKey.read(@client_key_data)
      end
      http
    end

    # Build a typed ApiError from a non-2xx response, preferring the Kubernetes
    # Status object {kind:"Status", reason, code, message}. The token VALUE is
    # never echoed: only the API server's own Status text is surfaced.
    def api_error(status, raw_body)
      reason = nil
      message = "Kubernetes API request failed: HTTP #{status}"
      parsed = begin
        JSON.parse(raw_body.to_s)
      rescue StandardError
        nil
      end
      if parsed.is_a?(Hash) && parsed["kind"] == "Status"
        reason = parsed["reason"]
        message = parsed["message"] unless parsed["message"].nil? || parsed["message"].to_s.empty?
      end
      ApiError.new(message, status: status, reason: reason)
    end

    def config_error(detail)
      MitosError.new(
        "invalid Kubernetes client configuration",
        code: "k8s_config_error",
        cause: detail,
        remediation: "Pass a valid kubeconfig: or run AgentRun(in_cluster: true) inside a pod."
      )
    end
  end

  # ApiError is raised for a non-2xx Kubernetes API response. +status+ is the
  # HTTP status (callers branch on 404 / 409, never the message) and +reason+ is
  # the Kubernetes Status reason (for example "NotFound", "AlreadyExists"). It is
  # a MitosError so a caller that only rescues Mitos::MitosError still catches it.
  class ApiError < MitosError
    attr_reader :reason

    def initialize(message, status:, reason: nil)
      super(
        message,
        code: "k8s_api_error",
        cause: reason.to_s,
        remediation: "Inspect the Kubernetes API server response and RBAC for this resource.",
        status: status
      )
      @reason = reason
    end
  end
end
